<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="turtlefied">
    <xacro:include filename="$(find turtlefied_pkg)/urdf/arm.xacro" />
    <!-- this is for ros2 simulation hardware control setup -->
    <xacro:include filename="$(find turtlefied_pkg)/urdf/modpi_sim.ros2_control.xacro"/>
    <xacro:include filename="$(find turtlefied_pkg)/urdf/_d435i.xacro" />

    <!-- Links -->
    <link name="base_footprint"/>

    <link name="base_link">
    <inertial>
        <mass value="1.3729096e+00"/>
        <inertia ixx="8.7002718e-03" ixy="-4.7576583e-05" ixz="1.1160499e-04"
                iyy="8.6195418e-03" iyz="-3.5422299e-06"
                izz="1.4612727e-02" />
    </inertial>
    <visual>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/modpi_base.stl" scale="0.01 0.01 0.01"/>
        </geometry>
        <material name="grey"/>
    </visual>
    <collision>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/modpi_base.stl" scale="0.01 0.01 0.01"/>
        </geometry>
    </collision>
    </link>

    <link name="wheel_left_link">
    <inertial>
        <mass value="2.8498940e-02" />
        <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
                iyy="1.1192413e-05" iyz="-1.4400107e-11"
                izz="2.0712558e-05" />
        </inertial>
    <visual>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/wheel.stl" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="dark"/>
    </visual>
    <collision>
        <geometry>
        <cylinder length="0.033" radius="0.055"/>
        </geometry>
    </collision>
    </link>

    <link name="wheel_right_link">
    <inertial>
        <mass value="2.8498940e-02" />
        <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
                iyy="1.1192413e-05" iyz="-1.4400107e-11"
                izz="2.0712558e-05" />
        </inertial>
    <visual>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/wheel.stl" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="dark"/>
    </visual>
    <collision>
        <geometry>
        <cylinder length="0.033" radius="0.055"/>
        </geometry>
    </collision>
    </link>

    <link name="roller_l_link">
    <inertial>
        <mass value="0.007" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0"
                izz="0.001" />
    </inertial>
    <visual>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/rollerframe.stl" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="light_black"/>
    </visual>
    <collision>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/rollerframe.stl" scale="0.001 0.001 0.001"/>
        </geometry>
    </collision>
    </link>

    <link name="roller_r_link">
    <inertial>
        <mass value="0.007" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0"
                izz="0.001" />
    </inertial>
    <visual>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/rollerframe.stl" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="light_black"/>
    </visual>
    <collision>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/rollerframe.stl" scale="0.001 0.001 0.001"/>
        </geometry>
    </collision>
    </link>

    <link name="imu_link"/>

    <link name="base_scan">
    <inertial>
        <mass value="0.114" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0"
                izz="0.001" />
    </inertial>
    <visual>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/lds.stl" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="dark"/>
    </visual>
    <collision>
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/lds.stl" scale="0.001 0.001 0.001"/>
        </geometry>
    </collision>
    </link>

    <link name="structure_link">
    <inertial>
        <mass value="0.035" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0"
                izz="0.001" />
    </inertial>
    <visual>
        <origin rpy="0 0 -1.57" />
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/st01.stl" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="blue"/>
    </visual>
    <collision>
        <origin rpy="0 0 -1.57" />
        <geometry>
        <mesh filename="package://turtlefied_pkg/meshes/st01.stl" scale="0.001 0.001 0.001"/>
        </geometry>
    </collision>
    </link>

    <link name="camera_depth_optical_frame"></link>

     <!-- Joints -->
    <joint name="base_joint" type="fixed">
        <parent link="base_footprint"/>
        <child link="base_link" />
        <origin xyz="0 0 0.04" rpy="0 0 0"/>
    </joint>

    <joint name="imu_joint" type="fixed">
        <parent link="base_link"/>
        <child link="imu_link"/>
        <origin xyz="0.032 0 0.29" rpy="0 0 0"/>
    </joint>

    <joint name="scan_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_scan"/>
        <origin xyz="0 0 0.43" rpy="0 0 0"/>
    </joint>

    <joint name="wheel_left_joint" type="continuous">
        <parent link="base_link"/>
        <child link="wheel_left_link"/>
        <origin xyz="0.064 0.144 0.01715" rpy="-1.57 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="wheel_right_joint" type="continuous">
        <parent link="base_link"/>
        <child link="wheel_right_link"/>
        <origin xyz="0.064 -0.144 0.01715" rpy="-1.57 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="roller_l_joint" type="fixed">
        <parent link="base_link"/>
        <child link="roller_l_link"/>
        <origin xyz="-0.086 0.064 0" rpy="0 0 1.57"/>
    </joint>

    <joint name="roller_r_joint" type="fixed">
        <parent link="base_link"/>
        <child link="roller_r_link"/>
        <origin xyz="-0.086 -0.064 0" rpy="0 0 1.57"/>
    </joint>

    <joint name="structure_joint" type="fixed">
        <parent link="base_link" />
        <child link="structure_link" />
        <origin xyz="0.11 0 0.06" rpy="0 -0.35 0" />
    </joint>

    <joint name="optical_joint" type="fixed">
        <parent link="structure_link" />
        <child link="camera_depth_optical_frame" />
        <origin xyz="0 0 0" rpy="-1.57 0 -1.57" />
    </joint>

    <xacro:robot_arm parent="base_link">
        <origin xyz="0 0 0.914" rpy="0 0 0" />
    </xacro:robot_arm>

    <xacro:arm_ros2_control name="RobotSystem"/>

    <xacro:sensor_d435i parent="tool_link">
        <origin xyz="0 0 0.06" rpy="0 -1.57 0"/>
    </xacro:sensor_d435i>

</robot>
